File: tests/fixtures/node-domain/src/payment/service.ts
import { PaymentRepository, PaymentFilter, ListResult } from "./repository.js";

/** Service orchestrates payment use cases. */
export class PaymentService {
  constructor(private readonly repo: PaymentRepository) {}

  async listPayments(filter: PaymentFilter): Promise<ListResult> {
    const sanitized: PaymentFilter = {
      ...filter,
      page: filter.page < 1 ? 1 : filter.page,
      pageSize:
        filter.pageSize < 1 || filter.pageSize > 100 ? 20 : filter.pageSize,
    };
    return this.repo.list(sanitized);
  }

  async confirmPayment(id: string): Promise<void> {
    const payment = await this.repo.findById(id);
    if (!payment) throw new Error("payment not found");
    payment.confirm();
    await this.repo.save(payment);
  }
}

File: tests/fixtures/python-domain/src/payment/service.py
"""Payment use cases."""

from __future__ import annotations

from .entity import Payment
from .repository import PaymentFilter, PaymentRepository, ListResult


class PaymentService:
    """Orchestrates payment use cases."""

    def __init__(self, repo: PaymentRepository) -> None:
        if repo is None:
            raise ValueError("repository is required")
        self._repo = repo

    def list_payments(self, filter_: PaymentFilter) -> ListResult:
        page = filter_.page if filter_.page >= 1 else 1
        page_size = filter_.page_size if 1 <= filter_.page_size <= 100 else 20
        sanitized = PaymentFilter(
            customer_id=filter_.customer_id,
            status=filter_.status,
            page=page,
            page_size=page_size,
        )
        return self._repo.list_payments(sanitized)

    def confirm_payment(self, payment_id: str) -> None:
        payment = self._repo.find_by_id(payment_id)
        if payment is None:
            raise ValueError("payment not found")
        payment.confirm()
        self._repo.save(payment)
